### PR TITLE
The Autodownloader.

### PR DESCRIPTION
**Pull request by shrit**
This pull request provides the Autodownloader for mlpack dependencies.
It can be used with current and future dependencies.
This one is born from working with #2531, with help from @rcurtin and @zoq.

It is ready to merge, with minor reviews are always welcome.

### CODE REVIEW

**Review by zoq — APPROVED** [review]
Looks good to me, thanks for putting this together :+1:

File: CMakeLists.txt [hunk 1]
```diff
@@ -254,131 +260,67 @@ endif()
...
 #   STB_IMAGE_INCLUDE_DIR - include directory for STB image library
 #   MATHJAX_ROOT - root of MathJax installation
-find_package(Armadillo "${ARMADILLO_VERSION}" REQUIRED)
 
+if (DISABLE_DOWNLOADS)
+  find_package(Armadillo "${ARMADILLO_VERSION}" REQUIRED)
+else()
+  find_package(Armadillo "${ARMADILLO_VERSION}")
+  if (NOT ARMADILLO_FOUND)
+    get_deps(http://files.mlpack.org/armadillo-10.3.0.tar.gz armadillo armadillo-10.3.0.tar.gz)
```
Comment (Yashwants19) [inline CMakeLists.txt:273]: I am not sure, But I guess we can maintained a copy of the latest armadillo tarball(`armadillo-latest.tar.gz`) same as `ensmallen` at our website. Let me know what you think??

File: CMakeLists.txt [hunk 2]
```diff
@@ -254,131 +260,67 @@ endif()
...
+set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} "${ENSMALLEN_INCLUDE_DIR}")
 
-find_package(cereal "${CEREAL_VERSION}" REQUIRED)
+# Find cereal.
+if (DISABLE_DOWNLOADS)
+  find_package(cereal "${CEREAL_VERSION}" REQUIRED)
+else()
+  find_package(cereal "${CEREAL_VERSION}")
+  if (NOT CEREAL_FOUND)
+    get_deps(https://github.com/USCiLab/cereal/archive/refs/tags/v1.3.0.tar.gz cereal cereal-1.3.0.tar.gz)
```
Comment (Yashwants19) [inline CMakeLists.txt:319]: Can we some how fetch the latest version of cereal before downloading it???

File: CMakeLists.txt [hunk 3]
```diff
@@ -409,8 +351,15 @@ set(Boost_ADDITIONAL_VERSIONS
...
 # CMake'.  Good luck!  You'll need it.
 set(Boost_NO_BOOST_CMAKE 1)
-find_package(Boost "${BOOST_VERSION}")
-
+if (DISABLE_DOWNLOADS)
+  find_package(Boost "${BOOST_VERSION}" REQUIRED)
+else()
+  find_package(Boost "${BOOST_VERSION}")
+  if (NOT Boost_FOUND)
+    get_deps(https://dl.bintray.com/boostorg/release/1.75.0/source/boost_1_75_0.tar.gz boost boost_1_75_0.tar.gz)
```
Comment (Yashwants19) [inline CMakeLists.txt:359]:
```suggestion
    list(GET Boost_ADDITIONAL_VERSIONS 0 Boost_LATEST_VERSION_1)
    string(REPLACE "." "_" Boost_LATEST_VERSION_2 ${Boost_LATEST_VERSION_1})
    get_deps(https://dl.bintray.com/boostorg/release/${Boost_LATEST_VERSION_1}/source/boost_${Boost_LATEST_VERSION_2}.tar.gz boost boost_${Boost_LATEST_VERSION_2}.tar.gz)
```

As latest version of boost is maintained by actions, we can rely on `Boost_ADDITIONAL_VERSIONS`, and Hence we can get rid of updating the version of boost regularly here. Let me know what you think???

**Review by Yashwants19 — COMMENTED** [review]
Ahh, I guess I am late. :sweat_smile:

**Comment by rcurtin** [comment]
> Ahh, I guess I am late. 😅

Sorry, I guess I hit merge too fast 😄

**Comment by shrit** [comment]
@rcurtin @Yashwants19 no worries I can open another pull request to change later to the latest version for cereal and armadillo, but we need to host the last on mlpack.org.
Considering boost, frankly, it does not worth the modifications since it will be removed in a couple of months.
Let me know what do you think

**Pull request merged**